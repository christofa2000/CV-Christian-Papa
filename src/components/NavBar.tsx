"use client";

import { motion } from "framer-motion";
import { Menu, X } from "lucide-react";
import { useEffect, useState } from "react";

const navItems = [
  { name: "Sobre mÃ­", href: "#sobre-mi" },
  { name: "Proyectos", href: "#proyectos" },
  { name: "Contacto", href: "#contacto" },
];

export default function NavBar() {
  const [activeSection, setActiveSection] = useState("");
  const [isMobileMenuOpen, setIsMobileMenuOpen] = useState(false);

  useEffect(() => {
    const observer = new IntersectionObserver(
      (entries) => {
        entries.forEach((entry) => {
          if (entry.isIntersecting) {
            setActiveSection(entry.target.id);
          }
        });
      },
      { threshold: 0.3 }
    );

    // Observe all sections
    navItems.forEach((item) => {
      const element = document.querySelector(item.href);
      if (element) {
        observer.observe(element);
      }
    });

    return () => observer.disconnect();
  }, []);

  const scrollToSection = (href: string) => {
    const element = document.querySelector(href);
    if (element) {
      element.scrollIntoView({ behavior: "smooth" });
    }
    setIsMobileMenuOpen(false);
  };

  return (
    <>
      {/* Skip Link */}
      <a
        href="#content"
        className="absolute -top-10 left-4 z-50 bg-indigo-600 text-white px-4 py-2 rounded-md transition-transform focus-ring"
        onFocus={(e) => e.target.scrollIntoView({ behavior: "smooth" })}
      >
        Saltar al contenido principal
      </a>

      {/* NavBar */}
      <motion.nav
        initial={{ y: -100 }}
        animate={{ y: 0 }}
        transition={{ duration: 0.5 }}
        className="fixed top-0 left-0 right-0 z-40 bg-neutral-950/80 backdrop-blur-md border-b border-white/10"
      >
        <div className="container mx-auto px-4">
          <div className="flex items-center justify-between h-16">
            {/* Logo */}
            <motion.div
              whileHover={{ scale: 1.05 }}
              className="font-semibold text-xl text-neutral-200"
            >
              Christian Papa
            </motion.div>

            {/* Desktop Menu */}
            <div className="hidden md:flex items-center space-x-8">
              {navItems.map((item) => (
                <button
                  key={item.name}
                  onClick={() => scrollToSection(item.href)}
                  className={`relative px-3 py-2 text-sm font-medium transition-colors focus-ring ${
                    activeSection === item.href.replace("#", "")
                      ? "text-indigo-400"
                      : "text-neutral-300 hover:text-neutral-200"
                  }`}
                >
                  {item.name}
                  {activeSection === item.href.replace("#", "") && (
                    <motion.div
                      layoutId="activeIndicator"
                      className="absolute bottom-0 left-0 right-0 h-0.5 bg-indigo-400"
                    />
                  )}
                </button>
              ))}
            </div>

            {/* Mobile Menu Button */}
            <button
              onClick={() => setIsMobileMenuOpen(!isMobileMenuOpen)}
              className="md:hidden p-2 focus-ring text-neutral-300 hover:text-neutral-200"
              aria-expanded={isMobileMenuOpen}
              aria-controls="mobile-menu"
              aria-label="Toggle mobile menu"
            >
              {isMobileMenuOpen ? <X size={24} /> : <Menu size={24} />}
            </button>
          </div>
        </div>

        {/* Mobile Menu */}
        <motion.div
          id="mobile-menu"
          initial={false}
          animate={{
            height: isMobileMenuOpen ? "auto" : 0,
            opacity: isMobileMenuOpen ? 1 : 0,
          }}
          transition={{ duration: 0.3 }}
          className="md:hidden overflow-hidden bg-neutral-900 border-t border-white/10"
        >
          <div className="px-4 py-4 space-y-2">
            {navItems.map((item) => (
              <button
                key={item.name}
                onClick={() => scrollToSection(item.href)}
                className={`block w-full text-left px-3 py-2 text-sm font-medium rounded-md transition-colors focus-ring ${
                  activeSection === item.href.replace("#", "")
                    ? "text-indigo-400 bg-indigo-500/10"
                    : "text-neutral-300 hover:text-neutral-200 hover:bg-neutral-800"
                }`}
              >
                {item.name}
              </button>
            ))}
          </div>
        </motion.div>
      </motion.nav>
    </>
  );
}
