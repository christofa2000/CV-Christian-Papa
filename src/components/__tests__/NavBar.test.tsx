import { render, screen } from "@testing-library/react";
import NavBar from "../NavBar";

describe("NavBar component", () => {
  it("renderiza el skip link para accesibilidad", () => {
    render(<NavBar />);
    expect(
      screen.getByText(/Saltar al contenido principal/i)
    ).toBeInTheDocument();
  });

  it("renderiza los enlaces de navegación", () => {
    render(<NavBar />);
    expect(screen.getAllByText(/Sobre mí/i).length).toBeGreaterThan(0);
    expect(screen.getAllByText(/Proyectos/i).length).toBeGreaterThan(0);
    expect(screen.getAllByText(/Contacto/i).length).toBeGreaterThan(0);
  });

  it("renderiza el botón de menú móvil", () => {
    render(<NavBar />);
    const mobileMenuButton = screen.getByLabelText(/Abrir\/Cerrar menú móvil/i);
    expect(mobileMenuButton).toBeInTheDocument();
  });

  it("tiene los atributos ARIA correctos en el botón móvil", () => {
    render(<NavBar />);
    const mobileMenuButton = screen.getByLabelText(/Abrir\/Cerrar menú móvil/i);
    expect(mobileMenuButton).toHaveAttribute("aria-expanded", "false");
    expect(mobileMenuButton).toHaveAttribute("aria-controls", "mobile-menu");
  });

  it("coincide con el snapshot", () => {
    const { container } = render(<NavBar />);
    expect(container).toMatchSnapshot();
  });
});
