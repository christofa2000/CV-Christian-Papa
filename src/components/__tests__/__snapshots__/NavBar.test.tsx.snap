// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`NavBar component coincide con el snapshot 1`] = `
<div>
  <a
    class="sr-only focus:not-sr-only focus:absolute focus:-top-10 focus:left-4 focus:z-[100] focus:bg-indigo-600 focus:text-white focus:px-4 focus:py-2 focus:rounded-md focus:transition-transform focus-ring"
    href="#content"
  >
    Saltar al contenido principal
  </a>
  <nav
    class="relative w-full bg-transparent"
  >
    <div
      class="mx-auto max-w-6xl md:max-w-7xl px-4 sm:px-6 lg:px-8"
    >
      <div
        class="h-16 flex items-center justify-center"
      >
        <div
          class="hidden md:flex items-center gap-6"
        >
          <button
            class="px-5 py-2 text-sm font-medium rounded-full border transition-colors focus-ring
                      text-neutral-300 hover:text-neutral-100 border-white/10 hover:bg-white/10"
          >
            Sobre mí
          </button>
          <button
            class="px-5 py-2 text-sm font-medium rounded-full border transition-colors focus-ring
                      text-neutral-300 hover:text-neutral-100 border-white/10 hover:bg-white/10"
          >
            Proyectos
          </button>
          <button
            class="px-5 py-2 text-sm font-medium rounded-full border transition-colors focus-ring
                      text-neutral-300 hover:text-neutral-100 border-white/10 hover:bg-white/10"
          >
            Contacto
          </button>
        </div>
        <button
          aria-controls="mobile-menu"
          aria-expanded="false"
          aria-label="Abrir/Cerrar menú móvil"
          class="md:hidden p-2 focus-ring text-neutral-300 hover:text-neutral-200"
        >
          <svg
            class="lucide lucide-menu"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="4"
              x2="20"
              y1="12"
              y2="12"
            />
            <line
              x1="4"
              x2="20"
              y1="6"
              y2="6"
            />
            <line
              x1="4"
              x2="20"
              y1="18"
              y2="18"
            />
          </svg>
        </button>
      </div>
      <div
        class="md:hidden overflow-hidden bg-neutral-950/70 backdrop-blur-sm border-t border-white/10"
        id="mobile-menu"
      >
        <div
          class="px-4 py-4 space-y-2"
        >
          <button
            class="block w-full text-left px-5 py-2 text-sm font-medium rounded-full transition-colors focus-ring border
                      text-neutral-300 hover:text-neutral-100 hover:bg-neutral-800/60 border-white/10"
          >
            Sobre mí
          </button>
          <button
            class="block w-full text-left px-5 py-2 text-sm font-medium rounded-full transition-colors focus-ring border
                      text-neutral-300 hover:text-neutral-100 hover:bg-neutral-800/60 border-white/10"
          >
            Proyectos
          </button>
          <button
            class="block w-full text-left px-5 py-2 text-sm font-medium rounded-full transition-colors focus-ring border
                      text-neutral-300 hover:text-neutral-100 hover:bg-neutral-800/60 border-white/10"
          >
            Contacto
          </button>
        </div>
      </div>
    </div>
  </nav>
</div>
`;
